src/coreclr/nativeaot/Runtime/interoplibinterface_objc.cpp

```diff
@@ -0,0 +1,187 @@
+    {
+        void* fn = obj->get_EEType()
+                      ->GetTypeManagerPtr()
+                      ->AsTypeManager()
+                      ->GetClasslibFunction(ClasslibFunctionId::ObjectiveCMarshalTryGetTaggedMemory);
+        
+        if (fn == nullptr)
+            return false;
+        
+        auto pTryGetTaggedMemoryCallback = reinterpret_cast<ObjCMarshalNative::TryGetTaggedMemoryCallback>(fn);
+
+        // It is illegal for any of the callouts to trigger a GC.
+        Thread * pThread = ThreadStore::GetCurrentThread();
+        pThread->SetDoNotTriggerGc();
+
+        // Due to the above we have better suppress GC stress.
+        bool fGcStressWasSuppressed = pThread->IsSuppressGcStressSet();
+        if (!fGcStressWasSuppressed)
+            pThread->SetSuppressGcStress();
+
+        bool result = pTryGetTaggedMemoryCallback(obj, tagged);
+
+        // Revert GC stress mode if we changed it.
+        if (!fGcStressWasSuppressed)
+            pThread->ClearSuppressGcStress();
+
+        pThread->ClearDoNotTriggerGc();
+
+        return result;
+    }
+
+    template <typename T>
+    T TryGetCallbackViaClasslib(_In_ Object * object, _In_ ClasslibFunctionId id)
+    {
+        void* fn = object->get_EEType()
+                        ->GetTypeManagerPtr()
+                        ->AsTypeManager()
+                        ->GetClasslibFunction(id);
+
+        if (fn != nullptr)
+        {
+            fn = ((ObjCMarshalNative::TryGetCallback)fn)();
+        }
+
+        return (T)fn;
+    }
+}
+
+// One time startup initialization.
+bool ObjCMarshalNative::Initialize()
+{
+    s_sLock.Init(CrstObjectiveCMarshalCallouts, CRST_DEFAULT);
+
+    return true;
+}
+
+bool ObjCMarshalNative::RegisterBeginEndCallback(void * callback)
+{
+    _ASSERTE(callback != nullptr);
+
+     // We must be in Cooperative mode since we are setting callbacks that
+     // will be used during a GC and we want to ensure a GC isn't occurring
+     // while they are being set.
+     _ASSERTE(ThreadStore::GetCurrentThread()->IsCurrentThreadInCooperativeMode());
+    
+    GcBeginEndCallbacks * pCallback = new (nothrow) GcBeginEndCallbacks();
+    if (pCallback == NULL)
+        return false;
+    
+    pCallback->m_pCallbackFunction = (ObjCMarshalNative::BeginEndCallback)callback;
+
+    CrstHolder lh(&s_sLock);
+
+    pCallback->m_pNext = m_pBeginEndCallbacks;
+    m_pBeginEndCallbacks = pCallback;
+
+    return true;
+}
+
+bool ObjCMarshalNative::IsTrackedReference(_In_ Object * object, _Out_ bool* isReferenced)
+{
+    *isReferenced = false;
+
+    if (!object->GetGCSafeMethodTable()->IsTrackedReferenceWithFinalizer())
+        return false;
+    
+    auto pIsReferencedCallbackCallback = TryGetCallbackViaClasslib<ObjCMarshalNative::IsReferencedCallback>(
+        object, ClasslibFunctionId::ObjectiveCMarshalGetIsTrackedReferenceCallback);
+
+    if (pIsReferencedCallbackCallback == nullptr)
+        return false;
```

Comment:
This shouldn't be needed and can be converted into an assert. I think it is hiding a bigger logical flaw if it returns `false`.

Reply:
I switched this to an assert. I also changed a couple other places to asserts where it should not be possible for something to happen. There are also now some comments with reasoning on why or why not to assert based on the result of `TryGetTaggedMemory`.